README


This archive contains the distribution Padre-Plugin-PerlTidy,
version 0.16:

  Format perl files using Perl::Tidy

This software is copyright (c) 2010 by Patrick Donelanm, Brian Cassidy.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Revision history for Perl extension Padre::Plugin::PerlTidy.

0.16   2010.12.06
    - Restored missing POT/PO translation sources (Damyan Ivanov, AZAWAWI)

0.15   2010.10.29
    - Removed Test::NeedsDisplay dependency (AZAWAWI)

0.14   2010.10.27
    - Fixed t/01-menu.t to work under no DISPLAY (AZAWAWI)
    - t/00-load.t is magically injected by [@PadrePlugin] via [LoadTests] (AZAWAWI)

0.13   2010.10.25
    - Switched to Dist::Zilla (AZAWAWI)
    - Fixed t/00-load.t to work under no DISPLAY (AZAWAWI)
    - Fix the uninitialized variable warning when tidying an unsaved tab (AZAWAWI)
    - Updated Arabic translation (AZAWAWI)

0.12  Tue Oct 12 2010
    - Fix duplicate "my $output" call

0.11  Mon Oct 11 2010
    - Added translations:
      - Turkish (BURAK)
    - When tidying and exporting an unsaved document or text, the tab title is
      now used instead of the undefined document filename (AZAWAWI)
    - Prevent the current destructive nature when tidying a non-Perl document
      or text (AZAWAWI)
    - Re-used standard Padre's message and error method calls instead of
      Wx::MessageBox (AZAWAWI, idea from SZABGAB++)
    - Fixed RT #61993 "Can't tidy code when the bottom pane is off on Padre startup"
      (AZAWAWI, Reported by RENEEB++)

0.10  Thu Jan 21 2010
    - A variety of cleanups to use Padre::Current more efficiently (ADAMK)
    - Added menu separator (ADAMK)
    - A somewhat cleaner code style that is closer to general conventions (ADAMK)
    - Added project config file that matches that of Padre itself (ADAMK)
    - Applied the Perl::Tidy plugin to itself, using the project config (ADAMK, BRICAS)
    - Added a few more menu tests (BRICAS)

0.09  Thu Jan 14 2010
    - Fixed #776: PerlTidy needs -nse and -nst options (MRWHIPPLE)
    - _T() replaced with Wx::gettext() (BRICAS)
    - Use new project config setting (config_perltidy) if available (BRICAS)

0.08  Mon Aug 17 2009
    - Center editor on active line after format (claudio)
    - Make cursor location storage less resource hungry (claudio)
    - Actually install the translations (RSN)
    - Added translations:
      - Japanese (ishigaki)
    - Updates for latest (0.43) API (BRICAS)
    - Add pod stubs so pod coverage will pass (BRICAS)
    - Renamed to the new Padre 0.43 PO naming scheme (azawawi)

0.07  Fri Jun 12 2009
    - Added translations:
      - German (hjansen)
      - Arabic (azawawi)
      - Spanish (brunov)
      - Brazilian Portuguese (garu)
    - Cursor location is kept after cleaning (instead of beginning document).
      (claudio)
    - Menu tests are now locale agnostic (BRICAS)

0.06  Wed May 20 2009
    - Keyboard shortcuts added (claudio)
    - Added translations:
      - French (jquelin)
      - Polish (therek)
      - Chinese (Traditional) translation (BLUET)

0.05  Mon Feb 09 2009
    - Use Padre 0.26 API for output window (KAARE)
    - Move author tests to release time checking (ADAMK + BRICAS)

0.04  Tue Jan 06 2009
    - Add HTML export feature (HJANSEN)
    - Capture errors from PerlTidy (HJANSEN)
    - Update for latest (0.23) API

0.03  Mon Dec 15 2008
    - Update for latest API

0.02  Thu Nov 27 2008
    - Use Module::Install::PadrePlugin to allow the user to install the plugin
      to their local dir
    - instead of use() Perl::Tidy, require() when needed
    - Update for the latest plugin system

0.01  Thu Oct 23 2008
    - original version


lib/Padre/Plugin/PerlTidy.pm
package Padre::Plugin::PerlTidy;
BEGIN {
  $Padre::Plugin::PerlTidy::VERSION = '0.15';
}

# ABSTRACT: Format perl files using Perl::Tidy


use 5.008002;
use strict;
use warnings;
use Params::Util   ();
use Padre::Current ();
use Padre::Wx      ();
use Padre::Plugin  ();

our @ISA = 'Padre::Plugin';

# This constant is used when storing
# and restoring the cursor position.
# Keep it small to limit resource use.
use constant {
	SELECTIONSIZE => 40,
};

sub padre_interfaces {
	'Padre::Plugin' => '0.43', 'Padre::Config' => '0.54';
}

sub plugin_name {
	Wx::gettext('Perl Tidy');
}

sub menu_plugins_simple {
	my $self = shift;
	return $self->plugin_name => [
		Wx::gettext("Tidy the active document\tAlt+Shift+F") => \&tidy_document,
		Wx::gettext("Tidy the selected text\tAlt+Shift+G") =>
			\&tidy_selection,
		'---' => undef,
		Wx::gettext('Export active document to HTML file') =>
			\&export_document,
		Wx::gettext('Export selected text to HTML file') =>
			\&export_selection,
	];
}

sub _tidy {
	my $main     = shift;
	my $current  = shift;
	my $source   = shift;
	my $document = $current->document;

	# Check for problems
	unless ( defined $source ) {
		return;
	}
	unless ( $document->isa('Padre::Document::Perl') ) {
		$main->error( Wx::gettext('Document is not a Perl document') );
		return;
	}

	my $destination = undef;
	my $errorfile   = undef;
	my %tidyargs    = (
		argv        => \'-nse -nst',
		source      => \$source,
		destination => \$destination,
		errorfile   => \$errorfile,
	);

	#Make sure output is visible...
	$main->show_output(1);
	my $output = $main->output;

	my $perltidyrc = $document->project->config->config_perltidy;
	if ($perltidyrc) {
		$tidyargs{perltidyrc} = $perltidyrc;
		$output->AppendText("Perl::Tidy running with project configuration $perltidyrc\n");
	} else {
		$output->AppendText("Perl::Tidy running with default or user configuration\n");
	}

	# TODO: suppress the senseless warning from PerlTidy
	require Perl::Tidy;
	eval { Perl::Tidy::perltidy(%tidyargs); };

	if ($@) {
		$main->error( Wx::gettext("PerlTidy Error") . ":\n" . $@ );
		return;
	}

	if ( defined $errorfile ) {
		my $filename = $document->filename ? $document->filename : $document->get_title;
		my $width = length($filename) + 2;
		$output->AppendText( "\n\n" . "-" x $width . "\n" . $filename . "\n" . "-" x $width . "\n" );
		$output->AppendText("$errorfile\n");
	}

	return $destination;
}

sub tidy_selection {
	my $main = shift;

	# Tidy the current selected text
	my $current = $main->current;
	my $text    = $current->text;
	my $tidy    = _tidy( $main, $current, $text );
	unless ( defined Params::Util::_STRING($tidy) ) {
		return;
	}

	# If the selected text does not have a newline at the end,
	# trim off any that Perl::Tidy has added.
	unless ( $text =~ /\n\z/ ) {
		$tidy =~ s{\n\z}{};
	}

	# Overwrite the selected text
	$current->editor->ReplaceSelection($tidy);
}

sub tidy_document {
	my $main = shift;

	# Tidy the entire current document
	my $current  = $main->current;
	my $document = $current->document;
	my $text     = $document->text_get;
	my $tidy     = _tidy( $main, $current, $text );
	unless ( defined Params::Util::_STRING($tidy) ) {
		return;
	}

	# Overwrite the entire document
	my ( $regex, $start ) = _store_cursor_position($current);
	$document->text_set($tidy);
	_restore_cursor_position( $current, $regex, $start );
}

sub _get_filename {
	my $main = shift;

	my $doc         = $main->current->document or return;
	my $current     = $doc->filename;
	my $default_dir = '';

	if ( defined $current ) {
		require File::Basename;
		$default_dir = File::Basename::dirname($current);
	}

	require File::Spec;

	while (1) {
		my $dialog = Wx::FileDialog->new(
			$main, Wx::gettext("Save file as..."),
			$default_dir, ( $current or $doc->get_title ) . '.html',
			"*.*", Wx::wxFD_SAVE,
		);
		if ( $dialog->ShowModal == Wx::wxID_CANCEL ) {
			return;
		}
		my $filename = $dialog->GetFilename;
		$default_dir = $dialog->GetDirectory;
		my $path = File::Spec->catfile( $default_dir, $filename );
		if ( -e $path ) {
			return $path if $main->yes_no( Wx::gettext("File already exists. Overwrite it?"), Wx::gettext("Exist") );
		} else {
			return $path;
		}
	}
}

sub _export {
	my ( $main, $src ) = @_;

	require Perl::Tidy;

	return unless defined $src;

	my $doc = $main->current->document;

	if ( !$doc->isa('Padre::Document::Perl') ) {
		$main->error( Wx::gettext('Document is not a Perl document') );
		return;
	}

	my $filename = _get_filename($main);

	return unless defined $filename;

	my ( $output, $error );
	my %tidyargs = (
		argv        => \'-html -nnn -nse -nst',
		source      => \$src,
		destination => $filename,
		errorfile   => \$error,
	);

	# Make sure output window is visible...
	$main->show_output(1);
	$output = $main->output;

	if ( my $tidyrc = $doc->project->config->config_perltidy ) {
		$tidyargs{perltidyrc} = $tidyrc;
		$output->AppendText("Perl\::Tidy running with project-specific configuration $tidyrc\n");
	} else {
		$output->AppendText("Perl::Tidy running with default or user configuration\n");
	}

	# TODO: suppress the senseless warning from PerlTidy
	eval { Perl::Tidy::perltidy(%tidyargs); };

	if ($@) {
		$main->error( Wx::gettext("PerlTidy Error") . ":\n" . $@ );
		return;
	}

	if ( defined $error ) {
		my $width = length( $doc->filename ) + 2;
		my $main  = Padre::Current->main;

		$output->AppendText( "\n\n" . "-" x $width . "\n" . $doc->filename . "\n" . "-" x $width . "\n" );
		$output->AppendText("$error\n");
		$main->show_output(1);
	}

	return;
}

sub export_selection {
	my $main = shift;
	my $text = $main->current->text;
	_export( $main, $text );
	return;
}

sub export_document {
	my $main = shift;
	my $text = $main->current->document->text_get;
	_export( $main, $text );
	return;
}

# parameter: $main, compiled regex
sub _restore_cursor_position {
	my $current = shift;
	my $regex   = shift;
	my $start   = shift;
	my $editor  = $current->editor;
	my $text    = $editor->GetTextRange(
		( $start - SELECTIONSIZE ) > 0 ? $start - SELECTIONSIZE
		: 0,
		( $start + SELECTIONSIZE < $editor->GetLength ) ? $start + SELECTIONSIZE
		: $editor->GetLength
	);
	eval {
		if ( $text =~ /($regex)/ )
		{
			my $pos = $start + length $1;
			$editor->SetCurrentPos($pos);
			$editor->SetSelection( $pos, $pos );
		}
	};
	$editor->goto_line_centerize( $editor->GetCurrentLine );
	return;
}

# parameter: $current
# returns: compiled regex, start position
# compiled regex is /^./ if no valid regex can be reconstructed.
sub _store_cursor_position {
	my $current = shift;
	my $editor  = $current->editor;
	my $pos     = $editor->GetCurrentPos;

	my $start;
	if ( ( $pos - SELECTIONSIZE ) > 0 ) {
		$start = $pos - SELECTIONSIZE;
	} else {
		$start = 0;
	}

	my $prefix = $editor->GetTextRange( $start, $pos );
	my $regex;
	eval {

		# Escape non-word chars
		$prefix =~ s/(\W)/\\$1/gm;

		# Replace whitespace by regex \s+
		$prefix =~ s/(\\\s+)/(\\s+|\\r*\\n)*/gm;

		$regex = qr{$prefix};
	};
	if ($@) {
		$regex = qw{^.};
		print STDERR @_;
	}
	return ( $regex, $start );
}

1;


__END__
=pod

=head1 NAME

Padre::Plugin::PerlTidy - Format perl files using Perl::Tidy

=head1 VERSION

version 0.15

=head1 SYNOPSIS

This is a simple plugin to run Perl::Tidy on your source code.

Currently there are no customisable options (since the Padre plugin system
doesn't support that yet) - however Perl::Tidy will use your normal .perltidyrc
file if it exists (see Perl::Tidy documentation).

=head1 INSTALLATION

You can install this module like any other Perl module and it will
become available in your Padre editor. However, you can also
choose to install it into your user's Padre configuration directory only:

=over 4

=item * Install the prerequisite modules.

=item * perl Makefile.PL

=item * make

=item * make installplugin

=back

This will install the plugin as PerlTidy.par into your user's ~/.padre/plugins
directory.

Similarly, "make plugin" will just create the PerlTidy.par which you can
then copy manually.

=head1 METHODS

=head2 padre_interfaces

Indicates our compatibility with Padre.

=head2 plugin_name

A simple accessor for the name of the plugin.

=head2 menu_plugins_simple

Menu items for this plugin.

=head2 tidy_document

Runs Perl::Tidy on the current document.

=head2 export_document

Export the current document as html.

=head2 tidy_selection

Runs Perl::Tidy on the current code selection.

=head2 export_selection

Export the current code selection as html.

=head1 AUTHORS

=over 4

=item *

Brian Cassidy <bricas@cpan.org>

=item *

Patrick Donelan

=item *

Ahmad M. Zawawi <ahmad.zawawi@gmail.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Patrick Donelanm, Brian Cassidy.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  'build_requires' => {
    'Locale::Msgfmt' => '0.15',
    'Module::Build' => '0.3601',
    'Test::More' => '0'
  },
  'configure_requires' => {
    'Module::Build' => '0.3601'
  },
  'dist_abstract' => 'Format perl files using Perl::Tidy',
  'dist_author' => [
    'Brian Cassidy <bricas@cpan.org>',
    'Patrick Donelan',
    'Ahmad M. Zawawi <ahmad.zawawi@gmail.com>'
  ],
  'dist_name' => 'Padre-Plugin-PerlTidy',
  'dist_version' => '0.16',
  'license' => 'perl',
  'module_name' => 'Padre::Plugin::PerlTidy',
  'recommends' => {},
  'recursive_test_files' => 1,
  'requires' => {
    'Padre' => '0.57',
    'Perl::Tidy' => '0'
  },
  'script_files' => [],
  'share_dir' => {
    'dist' => 'share'
  }
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
